stl/inc/ranges

```diff
@@ -6352,6 +6352,140 @@ namespace ranges {
+    };
+
+    template <class _Rng, class _Container, class... _Types>
+    concept _Constructible_and_insertable = constructible_from<_Container, _Types...>
+        && (_Can_push_back<_Container, range_reference_t<_Rng>>
+            || _Can_insert_end<_Container, range_reference_t<_Rng>>);
+
+    template <class _Reference, class _Container>
+    _NODISCARD constexpr auto _Container_inserter(_Container& _Cont) {
+        if constexpr (_Can_push_back<_Container, _Reference>) {
+            return back_insert_iterator{_Cont};
+        } else {
+            return insert_iterator{_Cont, _Cont.end()};
+        }
+    }
+
+    template <class _Container, input_range _Rng, class... _Types>
+        requires (!view<_Container>)
+    _NODISCARD constexpr _Container to(_Rng&& _Range, _Types&&... _Args) {
+        // clang-format on
+        constexpr bool _Ref_converts = convertible_to<range_reference_t<_Rng>, range_value_t<_Container>>;
+        if constexpr (_Ref_converts && constructible_from<_Container, _Rng, _Types...>) {
+            return _Container(_STD forward<_Rng>(_Range), _STD forward<_Types>(_Args)...);
+        } else if constexpr (_Ref_converts && constructible_from<_Container, from_range_t, _Rng, _Types...>) {
+            return _Container(from_range, _STD forward<_Rng>(_Range), _STD forward<_Types>(_Args)...);
+        } else if constexpr (_Ref_converts && _Common_constructible<_Rng, _Container, _Types...>) {
+            return _Container(_RANGES begin(_Range), _RANGES end(_Range), _STD forward<_Types...>(_Args)...);
+        } else if constexpr (_Ref_converts && _Constructible_and_insertable<_Rng, _Container, _Types...>) {
+            _Container _Cont(_STD forward<_Types>(_Args)...);
+            if constexpr (_Sized_and_reservable<_Rng, _Container>) {
+                _Cont.reserve(_RANGES size(_Range));
+            }
+            _RANGES copy(_Range, _Container_inserter<range_reference_t<_Rng>>(_Cont));
+            return _Cont;
+        } else if constexpr (input_range<range_reference_t<_Rng>>) {
+            const auto _Xform = [](auto&& _Elem) {
+                return _RANGES to<range_value_t<_Container>>(_STD forward<decltype(_Elem)>(_Elem));
+            };
+            return _RANGES to<_Container>(views::transform(_Range, _Xform), _STD forward<_Types>(_Args)...);
+        } else {
+            static_assert(_Always_false<_Container>, "the program is ill-formed per N4910 [range.utility.conv.to]/1.3");
+        }
+    }
+
+    template <class _Container>
+    struct _To_class_fn {
+        _STL_INTERNAL_STATIC_ASSERT(!view<_Container>);
+
+        template <input_range _Rng, class... _Types>
+        _NODISCARD constexpr auto operator()(_Rng&& _Range, _Types&&... _Args) const requires requires {
+            _RANGES to<_Container>(_STD forward<_Rng>(_Range), _STD forward<_Types>(_Args)...);
+        }
+        { return _RANGES to<_Container>(_STD forward<_Rng>(_Range), _STD forward<_Types>(_Args)...); }

```

Comment:
Shouldn't this be just `to<_Container>(...)`? Same below.

[\[range.utility.conv.adaptors\]/1.3][]:

> Its call pattern is `to<C>(r, bound_­args...)`, where `r` is the argument used in a function call expression of `f`.

[\[range.utility.conv.adaptors\]/1.3]: http://eel.is/c++draft/range.utility.conv.adaptors#1.3

Reply:
[contents]/3: "Whenever an unqualified name other than `swap` is used in the specification of a declaration `D` in [support] through [thread] or [depr], its meaning is established as-if by performing unqualified name lookup ([basic.lookup.unqual]) in the context of `D`."

Note that "unqualified name lookup" doesn't include ADL, so `to` in [range.utility.conv.adaptors] means `::std::ranges::to`. Our implementation conventionally avoids unwanted ADL for names by fully qualifying. (Wrapping the function-call expression in `()` so it's no longer an identifier-expression also suppresses ADL, so we could write `(to<_Container>)(...)` instead, but that's not the convention we use.)